File: src/interface_inforesults.pas
unit interface_inforesults;

interface

uses Global, interface_global;

type 
    rep_EventObsSim = Record
        Obsi : double;
        StdObsi : double;
        Simi : double;
        DDi : ShortInt;
        MMi : shortInt;
        YYYYi : INTEGER;
        end;

    rep_ArrayObsSim = ARRAY[1..100] OF rep_EventObsSim;

procedure StatisticAnalysis(
                    constref TypeObsSim : rep_TypeObsSim;
                    constref RangeObsMin, RangeObsMax : integer;
                    constref StrNr : string;
                    var Nobs : integer;
                    var ObsAver, SimAver, PearsonCoeff, RMSE : double;
                    var NRMSE, NScoeff, IndexAg : double;
                    var ArrayObsSim : rep_ArrayObsSim);

procedure StatisticAnalysis_wrap(
                    constref TypeObsSim : integer;
                    constref RangeObsMin, RangeObsMax : integer;
                    constref StrNr : PChar;
                    constref strlen : integer;
                    var Nobs : integer;
                    var ObsAver, SimAver, PearsonCoeff, RMSE : double;
                    var NRMSE, NScoeff, IndexAg : double;
                    var ArrayObsSim : rep_ArrayObsSim);
    external 'aquacrop' name '__ac_interface_inforesults_MOD_statisticanalysis_wrap';

procedure WriteAssessmentSimulation(constref StrNr, totalnameEvalStat : string;
                                    constref TheProjectType : repTypeProject;
                                    constref RangeMin, RangeMax : integer);

procedure WriteAssessmentSimulation_wrap(
                                    constref p : PChar;
                                    constref strlen : integer;
                                    constref TheProjectType : ShortInt;
                                    constref RangeMin, RangeMax : integer);
    external 'aquacrop' name '__ac_interface_inforesults_MOD_writeassessmentsimulation_wrap';


implementation


procedure StatisticAnalysis(
                    constref TypeObsSim : rep_TypeObsSim;
                    constref RangeObsMin, RangeObsMax : integer;
                    constref StrNr : string;
                    var Nobs : integer;
                    var ObsAver, SimAver, PearsonCoeff, RMSE : double;
                    var NRMSE, NScoeff, IndexAg : double;
                    var ArrayObsSim : rep_ArrayObsSim);
var
    int_typeObsSim : integer;
    p : PChar;
    strlen : integer;
begin
    int_typeObsSim := ord(TypeObsSim);
    p := PChar(StrNr);
    strlen := Length(StrNr);
    StatisticAnalysis_wrap(int_TypeObsSim, RangeObsMin, RangeObsMax, p, strlen, 
                           Nobs, ObsAver, SimAver, PearsonCoeff, RMSE, NRMSE,
                           NScoeff, IndexAg, ArrayObsSim);
end;


procedure WriteAssessmentSimulation(constref StrNr, totalnameEvalStat : string;
                                    constref TheProjectType : repTypeProject;
                                    constref RangeMin, RangeMax : integer);
var
    int_projecttype : ShortInt;
    p : PChar;
    strlen : integer;
begin
    int_projecttype := ord(TheProjectType);
    p := PChar(totalnameEvalStat);
    strlen := Length(totalnameEvalStat);
    WriteAssessmentSimulation_wrap(p, strlen, int_projecttype, RangeMin, RangeMax)
end;


initialization


end.

File: src/interface_inforesults.f90
module ac_interface_inforesults

use, intrinsic :: iso_c_binding, only: c_ptr

use ac_kinds, only: dp, &
                    intEnum, &
                    int32

use ac_interface_global, only:  pointer2string
use ac_inforesults, only:   StatisticAnalysis, &
                            rep_EventObsSim, &
                            WriteAssessmentSimulation

implicit none


contains

subroutine StatisticAnalysis_wrap(TypeObsSim, RangeObsMin, RangeObsMax, StrNr_ptr, &
                             strlen, Nobs, ObsAver, SimAver, PearsonCoeff, RMSE, &
                             NRMSE, NScoeff, IndexAg, ArrayObsSim)
    integer(intEnum), intent(in) :: TypeObsSim
    integer(int32), intent(in) :: RangeObsMin
    integer(int32), intent(in) :: RangeObsMax
    type(c_ptr), intent(in) :: StrNr_ptr
    integer(int32), intent(in) :: strlen
    integer(int32), intent(inout) :: Nobs
    real(dp), intent(inout) :: ObsAver
    real(dp), intent(inout) :: SimAver
    real(dp), intent(inout) :: PearsonCoeff
    real(dp), intent(inout) :: RMSE
    real(dp), intent(inout) :: NRMSE
    real(dp), intent(inout) :: NScoeff
    real(dp), intent(inout) :: IndexAg
    type(rep_EventObsSim), dimension(100), intent(inout) :: ArrayObsSim

    character(len=strlen) :: string

    string = pointer2string(StrNr_ptr, strlen)
    call StatisticAnalysis(TypeObsSim, RangeObsMin, RangeObsMax, string, &
                           Nobs, ObsAver, SimAver, PearsonCoeff, RMSE, &
                           NRMSE, NScoeff, IndexAg, ArrayObsSim)
end subroutine StatisticAnalysis_wrap


subroutine WriteAssessmentSimulation_wrap(p, strlen, &
                                          TheProjectType, RangeMin, RangeMax)
    type(c_ptr), intent(in) :: p
    integer(int32), intent(in) :: strlen
    integer(intEnum), intent(in) :: TheProjectType
    integer(int32), intent(in) :: RangeMin
    integer(int32), intent(in) :: RangeMax

    character(len=strlen), allocatable :: string

    string = pointer2string(p, strlen)

    call WriteAssessmentSimulation('', string, &
                                   TheProjectType, RangeMin, RangeMax)
end subroutine WriteAssessmentSimulation_wrap

end module ac_interface_inforesults
